aten/src/ATen/native/native_functions.yaml

```diff
@@ -4405,14 +4405,17 @@
 - func: is_pinned(Tensor self, Device? device=None) -> bool
   variants: method
   dispatch:
-    CUDA: is_pinned_cuda
+    NestedTensorCUDA, CUDA: is_pinned_cuda
     MPS: is_pinned_mps
     CompositeExplicitAutograd: is_pinned_default
 
 # TODO: add a copy kwarg that guarantees that the tensor is put into fresh
 # pinned memory
 - func: pin_memory(Tensor(a) self, Device? device=None) -> Tensor(a)
   variants: method
+  dispatch:
+    NestedTensorCUDA, NestedTensorCPU: pin_memory_nested
```

Comment:
You should use _pin_memory here to re-use the compositeimplicit part like we do for other devices.

Reply:
I tried this initially but it doesn't quite work since the `at::_pin_memory()` call is dispatched to this impl first:
https://github.com/pytorch/pytorch/blob/4069d1de59bb5cd60c5364acd2835fbf9c6e601c/aten/src/ATen/templates/RegisterBackendSelect.cpp#L36-L40

For the NT CUDA case, this wrongly results in dispatching to `_pin_memory_cuda` instead of `_pin_memory_nested`. I opted to impl `pin_memory` for NT instead to avoid this, but I could probably be convinced to tweak the above `_pin_memory` instead.

Reply:
> Why not add logic to this backend select kernel to properly handle the case where self is nested (like we need to do to support sparse as well)?

Sounds good, I think any overhead for this check should be negligible enough.